### PR TITLE
extend/ENV/super: allow `O{1,0}` to accept a block

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This makes `ENV.O{1,0}` behave like `ENV.deparallelize`.

This should also allow us to build libgcrypt's jitter entropy collector,
which we currently disable because it interacts poorly with our compiler
shims. See #11201 for a previous attempt at this.

### CODE REVIEW

File: Library/Homebrew/extend/ENV/super.rb [hunk 1]
```diff
@@ -338,8 +338,19 @@ def refurbish_args
...
 
   %w[O1 O0].each do |opt|
-    define_method opt do
+    define_method opt do |&block|
+      old = delete("HOMEBREW_OPTIMIZATION_LEVEL")
       send(:[]=, "HOMEBREW_OPTIMIZATION_LEVEL", opt)
+
+      if block
+        begin
+          block.call
```
Comment (carlocab) [inline Library/Homebrew/extend/ENV/super.rb:347]: Not sure why `yield` doesn't work, but I don't really get Ruby metaprogramming.

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me. I'd love to see more of these methods take block forms to be used in a more finely-grained manner (and perhaps even deprecate the non-block for of `deparallelize`). `with_env` may be a cleaner solution to this?

**Comment by carlocab** [comment]
> `with_env` may be a cleaner solution to this?

`with_env` would work if we change these methods to only accept blocks. I don't see a way of making it work if you want the environment change to persist until the end of the `install` method, but maybe I just don't understand `with_env`.

**Comment by MikeMcQuaid** [comment]
> `with_env` would work if we change these methods to only accept blocks. I don't see a way of making it work if you want the environment change to persist until the end of the `install` method, but maybe I just don't understand `with_env`.

I was thinking maybe just for the block case?

**Comment by carlocab** [comment]
So you mean something like
```ruby
  %w[O1 O0].each do |opt|
    define_method opt do |&block|
      if block
        with_env(HOMEBREW_OPTIMIZATION_LEVEL: opt) { block.call }
      else
        send(:[]=, "HOMEBREW_OPTIMIZATION_LEVEL", opt)
      end
    end
  end
```
?

**Comment by MikeMcQuaid** [comment]
@carlocab Yeh, that looks nice to me 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by carlocab** [comment]
Oops. Sorbet really does not like the new version:
```
❯ brew typecheck
extend/ENV/super.rb:343: This code is unreachable https://srb.help/7006
     343 |        with_env(HOMEBREW_OPTIMIZATION_LEVEL: opt) { block.call }
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    extend/ENV/super.rb:342: This condition was always falsy (NilClass)
     342 |      if block
                   ^^^^^
  Got NilClass originating from:
    extend/ENV/super.rb:341:
     341 |    define_method opt do |&block|
                                     ^^^^^
```

**Comment by MikeMcQuaid** [comment]
@carlocab Given there's now only two of these methods: I'd suggest avoiding `define_method` and defining them normally.

**Comment by carlocab** [comment]
Ah, just saw your comment. Convincing Sorbet that `block` can occasionally not be `nil` isn't too much more code. Would you still rather these two methods be defined separately?

**Comment by MikeMcQuaid** [comment]
@carlocab Sorry, yeh, I think it'd be nicer.

**Comment by carlocab** [comment]
Sure, that's fine. I've pushed the changes.

File: Library/Homebrew/extend/ENV/super.rb [hunk 2]
```diff
@@ -337,11 +337,25 @@ def refurbish_args
     append_to_cccfg "O"
   end
 
-  %w[O1 O0].each do |opt|
-    define_method opt do
-      send(:[]=, "HOMEBREW_OPTIMIZATION_LEVEL", opt)
+  # rubocop: disable Naming/MethodName
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/ENV/super.rb:340]: Can you add a comment here explaining why? Thanks!

**Review by MikeMcQuaid — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Thanks, sorry for all the back and forth 🙇🏻

**Comment by carlocab** [comment]
The apology is appreciate but very much unnecessary 🙂

**Pull request merged**